### ISSUE TITLE
Notes and considerations on "Why I like Django and why I like SAP" post

### ISSUE THREAD

**Issue by Suncatcher** [guettli/why-i-like-django-and-sap#1]
Hi,
I read your post about SAP/Python comparison, it looks very interesting, and I have couple of notes about it:

> Users can't create hyperlinks to the sap notes in Q+A sites

Of course they can. On Stackoverflow I give links to notes frequently like this https://launchpad.support.sap.com/#/notes/1151258

> because in the database a string like "00000001234" is stored

and here is the answer to your question, this is the **string**, not the integer. If you put leading zeroes in numeric literals

    WHERE netwr = 0001234

database perfectly eat it. Not sure how it works in Python, I believe direct queries to DB are not very common in Django and ORM tools are used.

> Therefore, the result of a logical expression cannot be assigned directly to a data object.

it can. You can write 

        DATA: bool TYPE abap_bool.

        bool = cl_ciap_rounding=>is_integer( 5 ).

and it will work. You can also use [predicative method call](https://stackoverflow.com/a/45813392/911419). What concrete tasks you cannot do with `abap_bool`?

> In SAP you almost always need custom data elements.

no. For what? You need to create custom data types only if you want they bear some textual info in GUI programs, for search-helps, etc. If you need just to save smth in table you can always use [elementary types](https://help.sap.com/doc/saphelp_nw70ehp3/7.03.19/en-US/fc/eb3138358411d1829f0000e829fbfe/content.htm?no_cache=true), they are plenty of them.

> I like Django because it is easy to use ForeignKeys which helps you to avoid redundancy. I was surprised that this is different in SAP.

yes, but this is because SAP data model is error-prone, unmaintainable and not starighforward. 
It doesn't mean one cannot use foreign key in his custom tables. The statement "SAP not uses foreign keys" makes one feel that there is no such concept in SAP at all, which is misleading.

> In my environment this takes long. Roughly 15 minutes.

It depends on the setup. I had a client where this process took seconds (automatic transports).

>  In ABAP you still store the datetime in two columns: One for the day, one for the time

[Timestamps](https://stackoverflow.com/questions/49053550/abap-how-to-add-days-to-a-date-of-type-timestamp) exist in ABAP since ancient times, as well as [tools for converting them](https://help.sap.com/doc/abapdocu_753_index_htm/7.53/en-US/abapconvert_time-stamp.htm). You can perfectly save values to the DB table with timestamp column type.

> You want to convert to create json from ABAP? Yes, it is possible. But there are too many ways.
> In Python it is easy (I love "no brainers"), since there is one lib in the standard library, no need to find the matching lib

Also I disagree about "single point of truth" in Python. If you search [pypi with json word](https://pypi.org/search/?q=json) you will find **more than 10.000 packages** for JSON tasks. But no one can oblige you to use them, you can do it with [standard lib](https://docs.python.org/3/library/json.html) or with custom one.

The same with ABAP json part. No one can make you to install 3rd-parties, you can just use [standard sXML library for JSON](https://help.sap.com/doc/abapdocu_753_index_htm/7.53/en-US/abenabap_sxml_lib.htm), as provided by SAP. If you don't like it, you are free to install custom frameworks.

There are many subjective and opinion-based points in your list, but nevertheless, it is nice you did such comparison. I've never read smth such comprehensive like yours 🙂

**Comment by guettli** [comment]
Hi Suncatcher, please create several issues the next time (not one big containing many issues). This makes it easier for me.

**Comment by guettli** [comment]
https://github.com/guettli/why-i-like-django-and-sap/commit/1393011628065b1313d7bce5b74841dc37afcd63
Solves "can't create hyperlink to docs".


"Therefore, the result of a logical expression cannot be assigned directly to a data object." ... this quote is from this page: https://help.sap.com/doc/abapdocu_752_index_htm/7.52/en-US/abendataobjects_true_value_guidl.htm
It this is not true, then the sap docs are not valid.
Is there a new version of above page which does not contain invalid statements?


> In SAP you almost always need custom data elements.
At least in the code which I have seen (SAP IS-U) a lot of custom data elements get used. In a normal SQL databases people would use a foreign-key. But in the code I have seen people seem to prefer to create custom data elements. But maybe only the inventors of SAP IS-U did structure their application like this.



You said:
> yes, but this is because SAP data model is error-prone, unmaintainable and not starighforward.
> It doesn't mean one cannot use foreign key in his custom tables. The statement "SAP not uses foreign keys" 
> makes one feel that there is no such concept in SAP at all, which is misleading.

Yes, you are right. I updated my page. 


> timestamp exists
Yes, you are right I updated the page.

About "json": There is one standard json lib in Python. And there are several other libraries which build upon this library. 

> There are many subjective and opinion-based points 
Yes, that is true. This page was like a barf bag when I wrote [modlink](https://www.tbz-pariv.de/produkte/modlink/uebersicht)

thank you suncatcher for your feedback!
